lib/queue.js

```diff
@@ -606,92 +666,103 @@ class Queue extends Emitter {
+            // This job will get picked up later as a stalled job if we happen
+            // to get here. We can't easily process this job because at this
+            // point Queue#close has already captured the activeJobs set in a
+            // Promise.all call, and we'd prefer to delay a job than
+            // half-process it.
+            this.queued -= 1;
+            return;
+          }
 
-          setImmediate(jobTick);
+          this.running += 1;
+          this.queued -= 1;
+          if (this.running + this.queued < this.concurrency) {
+            this.queued += 1;
+            setImmediate(jobTick);
+          }
 
-          /* istanbul ignore else */
-          if (results) {
-            const status = results[0], result = results[1];
-            this.emit(status, job, result);
+          if (!job) {
+            // Per comment in Queue#_waitForJob, this branch is possible when
+            // the job is removed before processing can take place, but after
+            // being initially acquired.
+            return;
           }
-        });
-      }, (err) => {
-        setImmediate(jobTick);
-        throw err;
-      }).catch((err) => this.emit('error', err));
+
+          return this._runJob(job).then((results) => {
+            this.running -= 1;
+            this.queued += 1;
+
+            /* istanbul ignore else */
+            if (results) {
+              const status = results[0],
+                result = results[1];
+              this.emit(status, job, result);
+            }
+          }, this._emitErrorAfterTick);
+        }),
+        () => setImmediate(jobTick)
+      ).catch(this._emitErrorAfterTick);
     };
 
-    this.checkStalledJobs(jobTick);
+    this._doStalledJobCheck().then(jobTick).catch(this._emitErrorAfterTick);
     this._activateDelayed();
 
     return this;
   }
 
   _doStalledJobCheck() {
-    return this._evalScript('checkStalledJobs', 4, this.toKey('stallBlock'),
-      this.toKey('stalling'), this.toKey('waiting'), this.toKey('active'),
-      this.settings.stallInterval)
-      .then((stalled) => {
-        for (let jobId of stalled) {
-          this.emit('stalled', jobId);
-        }
-        return stalled.length;
-      });
+    return this._evalScript(
+      'checkStalledJobs',
+      4,
+      this.toKey('stallBlock'),
+      this.toKey('stalling'),
+      this.toKey('waiting'),
+      this.toKey('active'),
+      this.settings.stallInterval
+    ).then((stalled) => {
+      for (let jobId of stalled) {
+        this.emit('stalled', jobId);
+      }
+      return stalled.length;
+    });
+  }
+
+  _safeCheckStalledJobs(interval, cb) {
+    const promise = this._checkStalledJobs(interval, cb);
+    // If a callback is not defined, then we must emit errors to avoid unhandled
+    // rejections. If there is a callback, then _checkStalledJobs will attach it
+    // as an error handler to `promise`.
+    if (!cb) promise.catch(this._emitErrorAfterTick);
```

Comment:
Any drawback to attaching this in `_checkStalledJobs` itself?

Reply:
Yeah - that'd result in the `Queue` emitting an `error` event even in the case that we're calling `checkStalledJobs` directly and handling the returned `Promise`. This only exists for subsequent calls on the originally specified interval.